File: src/main/java/at/ac/tuwien/complang/iterative/OptimizedNQueensSolverIterativeArray.java

package at.ac.tuwien.complang.iterative;

/**
 *
 */
public class OptimizedNQueensSolverIterativeArray {
    
    private final int n;
    private final int mid;
    
    // 1..n
    // a[x] ist true, wenn man in Spalte x eine Dame setzen kann
    private final boolean[] a;
    
    // 2..2*n
    // b[x+y] ist true, wenn man in der Diagonale, 
    // die die Felder (x+k, y-k) (fuer alle k) enthaelt, eine Dame setzen kann
    private final boolean[] b;
    
    // 1..2*n-1
    // c[x-y+n] ist true, wenn man in der Diagonale,
    // die die Felder (x+k, y+k) (fuer alle k) enthaelt, eine Dame setzen kann
    private final boolean[] c;
    
    // 1..n
    // x[y] enthaelt die Spalten der aktuellen (Teil)loesung
    private final int[] x;

    private long count = 0;
    
    public OptimizedNQueensSolverIterativeArray(int n) {
        this.n = n;
        this.mid = n >> 1;
        this.x = new int[n + 1];
        this.a = new boolean[n + 1];
        this.b = new boolean[2 * n + 1];
        this.c = new boolean[2 * n];
        int i;

        for (i = 1; i <= n; i++) {
            a[i] = true;
        }
        for (i = 2; i <= 2 * n; i++) {
            b[i] = true;
        }
        for (i = 1; i <= 2 * n - 1; i++) {
            c[i] = true;
        }
    }
    
    private long stackElement(int col, int row) {
        return ((long) col << 32) | (long) row;
    }
    
    private int col(long stackElement) {
        return (int) (stackElement >> 32);
    }
    
    private int row(long stackElement) {
        return (int) stackElement & -1;
    }

    /**
     * Do iterative DFS for every field in the first row.
     */
    private void try0() {
        /* Stack */
        final long[] stack = new long[n * n];
        
        int j;
        int jt;
        /* The currently processed field. */
        long f;
        int fRow;
        int fCol;
        /* Pointer to next free entry on top of Stack. If 0, the stack is empty. */
        int last = 0;
        
        // Don't count mirrored solutions, so only count to the middle
        int odd = n & 1;
        int outerEnd = mid + odd;
        int lastRow = -1;
    
        for (j = 1; j <= outerEnd; j++) {
            final int bIndex = 1 + j;
            final int cIndex = 1 - j + n;
            // do DFS for a queen placed in first (topmost) row on column j (1..n)
            // mark queen placed in field of first row (j,1)
            x[1] = j;
            a[j] = false;
            b[bIndex] = false;
            c[cIndex] = false;
            
            if (j == outerEnd && odd == 1) {
                // when n is odd, only count to the middle column on the second row
                for (jt = 1; jt <= mid; jt++) {
                    // push current field onto stack
                    stack[last++] = stackElement(jt, 2);
                }
                
            } else {
                // push all fields from row 2 on stack, so we can process them in a
                // dfs way.
                for (jt = 1; jt <= n; jt++) {
                    // push current field onto stack
                    stack[last++] = stackElement(jt, 2);
                }
            }
            // If stack is empty we have checked all edges in our search tree.
            while (last != 0) {
                // Grab next field from stack and do dfs on it.
                // For simple structs (as ours) the compiler will make memcopy, so
                // we can use simple assign here.
                f = stack[--last];
                fRow = row(f);
                fCol = col(f);
                if (fRow < lastRow) {
                    // There is already a queen placed in the current row. So we
                    // know, that we have gone backwards in the tree - and we need
                    // to clear the queen in the current row. And also in all rows
                    // beneath us. It might happen that there was no field from next
                    // row processed during backtracking - and then the
                    // queen will be still there in the next row!
                    clearQueens(fRow, lastRow);
                }
                if (isValid(fCol, fRow)) {
                    // place queen in this field
                    x[fRow] = fCol;
                    a[fCol] = false;
                    b[fRow + fCol] = false;
                    c[fRow - fCol + n] = false;
                    if (fRow != n) {
                        // We have to go 1 row deeper and push all fields on stack.
                        for (jt = 1; jt <= n; jt++) {
                            stack[last++] = stackElement(jt, fRow + 1);
                        }
                    } else {
                        // We hit last row, this is a valid solution.
                        count++;
                        // We also need to clean up board for the last row here!
                        a[fCol] = true;
                        b[fRow + fCol] = true;
                        c[fRow - fCol + n] = true;
                        //printSolution(); // TODO: skip this line for perf tests
                        x[fRow] = 0;
                        lastRow = fRow;
                    }
                } else {
                    lastRow = fRow;
                }
            }
            // Clear queen in first row. All other queens have been removed already.
            a[j] = true;
            b[bIndex] = true;
            c[cIndex] = true;
        }
    }
    
    /**
     * Return true only if a queen can savely be placed in column x and row y.
     * This is Niklaus Wirth's validation algorithm.
     */
    private boolean isValid(int x, int y) {
        return a[x] && b[y + x] && c[y - x + n];
    }

    /**
     * This clears all eventually leftover queens from the board beginning at
     * startRow downwards. We can skip last row, because that one will always be
     * cleared.
     */
    private void clearQueens(int startRow, int lastRow) {
        for (int y = startRow; y <= lastRow; y++) {
            if (x[y] != 0) {
                // clear Niklaus Wirth's arrays
                a[x[y]] = true;
                b[y + x[y]] = true;
                c[y - x[y] + n] = true;
                x[y] = 0;
            }
        }
    }

    public long solve() {
        try0();
        
        // Add mirrored solutions
        count = count << 1;
        
        return count;
    }

    public static void main(String[] args) {
        int n = Integer.parseInt(args[0]);
        OptimizedNQueensSolverIterative o = new OptimizedNQueensSolverIterative(n);
        System.out.println(o.solve());
    }
}

File: src/main/c/OptimizedNQueensSolverIterativeArray.c
/******************************************************************************
 * converted from recursive to interative program.
 * only clear queens from board if required
 * replace 'struct field f' (stack element) with simple integers
 * use char instead of integer in 'struct field' for stack elements
 * Symmetrie checking along the vertical middle line
 *
 *****************************************************************************/
#include <stdio.h>
#include <stdlib.h>

#define MAXN 31
/* this is the actual board size to be used */
int n;
/* half of n */
int mid;
/* number of found solutions to the n-queens problem */
long count = 0L;
/* a[x] is true, if a queen can be placed in column x  (x: 1..n) */
int *a;
/* b[y+x] is true, if a queen can be placed in the diagonal which is covered
 by the fields (x+k, y-k) for all k.
 "/" antidiagonal line) */
int *b;
/* c[y-x+n] is true, if a queen can be placed in the diagonal which is covered
 by the fields (x+k, y+k) for all k.
 "\" diagonal line */
int *c;
/* x[y] has an entry for each row which tells, in which column the queen is
 put for current partly solution. */
int *x;

struct field {
	char col;
	char row;
};

/* Stack */
struct field stack[MAXN * MAXN];

/* Pointer to next free entry on top of Stack. If 0, the stack is empty. */
int last = 0;

/*
 * Return true only if a queen can savely be placed in column x and row y.
 * This is Niklaus Wirth's validation algorithm.
 */
int isValid(int x, int y) {
	if (a[x] && b[y + x] && c[y - x + n]) {
		return 1;
	} else {
		return 0;
	}
}

/*
 * Do iterative DFS for every field in the first row.
 */
void try(void) {
	int j = 1; // main column counter (1..n)
	int jt = 1; // temporary column counter
	int y = 1; // start row for cleaning up queens
	int ccol = 1; // col number of current processed field
	int crow = 1; // row number of current processed field
	// Do not count mirrored solutions, so only count up to the middle column.
	int odd = n & 1;
	int outerEnd = mid + odd;

	for (j = 1; j <= outerEnd; j++) {
		// do DFS for a queen placed in first (topmost) row on column j (1..n)
		// mark queen placed in field of first row (j,1)
		x[1] = j;
		a[j] = 0;
		b[1 + j] = 0;
		c[1 - j + n] = 0;

		if (j == outerEnd && odd == 1) {
			// when n is odd, only count to the middle column on the second row
			for (jt = 1; jt <= mid; jt++) {
				// push current field onto stack
				stack[last].col = (char) jt;
				stack[last++].row = (char) 2;
			}
		} else {
			// push fields from row 2 on stack, so we can process them in a
			// dfs way.
			for (jt = 1; jt <= n; jt++) {
				// push current field onto stack
				stack[last].col = (char) jt;
				stack[last++].row = (char) 2;
			}
		}
		// If stack is empty we have checked all edges in our search tree.
		while (last) {
			// Grab next field from stack and do dfs on it.
			ccol = stack[--last].col;
			crow = stack[last].row;
			if (x[crow]) {
				// There is already a queen placed in the current row. So we
				// know, that we have gone backwards in the tree - and we need
				// to clear the queen in the current row. And also in next row
				// beneath us. It might happen that there was no field from next
				// row processed during backtracking - and then the
				// queen will be still there in the next row!
				a[x[crow]] = 1;
				b[crow + x[crow]] = 1;
				c[crow - x[crow] + n] = 1;
				x[crow] = 0;
				if (x[crow + 1]) {
					a[x[crow + 1]] = 1;
					b[crow + 1 + x[crow + 1]] = 1;
					c[crow + 1 - x[crow + 1] + n] = 1;
					x[crow + 1] = 0;
				}
			}
			if (isValid(ccol, crow)) {
				// place queen in this field
				x[crow] = ccol;
				a[ccol] = 0;
				b[crow + ccol] = 0;
				c[crow - ccol + n] = 0;
				if (crow != n) {
					// We have to go 1 row deeper and push all fields on stack.
					for (jt = 1; jt <= n; jt++) {
						stack[last].col = (char) jt;
						stack[last++].row = (char) (crow + 1);
					}
				} else {
					// We hit last row, this is a valid solution.
					count++;
					// We also need to clean up board for the last row here!
					a[ccol] = 1;
					b[crow + ccol] = 1;
					c[crow - ccol + n] = 1;
					x[crow] = 0;
				}
			}
		}
		// clear any leftover queens in all rows (except the last one)
		for (y = 1; y < n; y++) {
			if (x[y]) {
				a[x[y]] = 1;
				b[y + x[y]] = 1;
				c[y - x[y] + n] = 1;
				x[y] = 0;
			}
		}
	}
}

int main(int argc, char *argv[]) {
	if (argc <= 1 || (n = atoi(argv[1])) <= 0) {
		n = 8;
	}
	mid = n >> 1; // mid = n/2
	// initialize all arrays
	// make arrays the required size
	a = (int *) malloc((n + 1) * sizeof(int));
	b = (int *) malloc((2 * n + 1) * sizeof(int));
	c = (int *) malloc((2 * n) * sizeof(int));
	x = (int *) malloc((n + 1) * sizeof(int));
	// initialize all arrays, set all true
	int i = 1;
	for (i = 1; i <= n; i++) {
		a[i] = 1; // '|' column storage
	}
	for (i = 2; i <= (2 * n); i++) {
		b[i] = 1; // '/' diagonal storage
	}
	for (i = 1; i <= ((2 * n) - 1); i++) {
		c[i] = 1; // '\' diagonale storage
	}
	for (i = 0; i <= (n + 1); i++) {
		x[i] = 0;
	}
	try(); // start
	// add mirrored solutions (simply double count)
	count = count << 1;
	printf("%ld\n", count); // print result (number of solutions found)
	return 0;
}

